### PR TITLE
NPUW: Head/tail optimizations

### PR DESCRIPTION
**Pull request by dmatveev**
### Details:
 - This PR enables WS and DQ for head/tail
 - Some efficiency problems were fixed by introducing a host-side Gather (ON by default, can be turned OFF for evaluation purposes)

### Tickets:
 - E-139867

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp [hunk 1]
```diff
@@ -1569,12 +1570,27 @@ void Partitioner::optimize(const std::string& func_name) {
     {
         ov::npuw::patterns::opt::Context ctx;
         ctx.pmm_dims = cfg.get<::intel_npu::NPUW_PMM>();
+
+        ov::pass::GraphRewrite rewr;
+        rewr.add_matcher<ov::npuw::patterns::opt::DQParMMGQ>(std::ref(ctx));
+        rewr.add_matcher<ov::npuw::patterns::opt::DQDictGatherCWu>(std::ref(ctx));
+        rewr.add_matcher<ov::npuw::patterns::opt::DQDictGatherGQi>(std::ref(ctx));
+        rewr.add_matcher<ov::npuw::patterns::opt::HostGather>(std::ref(ctx));
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1578]: This one can be removed?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 2]
```diff
@@ -144,6 +184,187 @@ DQMatMulCWi::DQMatMulCWi() {
...
+        auto gather_c = std::make_shared<ov::op::v0::Constant>(ov::element::i32, ov::Shape{}, 0);
+        auto new_g = std::make_shared<ov::op::v8::Gather>(new_wi, matched_out_ids, gather_c);
+        matched_node_cvt->input(0).replace_source_output(new_g);
+
+        return true;  // root has changed
+    };
+    register_matcher(std::make_shared<opp::Matcher>(qcvtm, "DQDictGatherCWu"), std::move(callback));
+}
+
+HostGather::HostGather(Context::Ref ctx) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:275]: Comment is missing here

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 3]
```diff
@@ -144,6 +184,187 @@ DQMatMulCWi::DQMatMulCWi() {
...
+            auto new_wi = ctx.get().unpack(matched_qweight, matched_qzerop, matched_qcoeff, ov::element::f16);
+            auto new_mm = std::make_shared<ov::op::v0::MatMul>(new_cvt_a, new_wi, false, true);
+            auto new_out = std::make_shared<ov::op::v0::Convert>(new_mm, ov::element::f32);
+
+            matched_result->input(0).replace_source_output(new_out);
+        }
+        return false;  // root has changed (yet)
+    };
+    register_matcher(std::make_shared<opp::Matcher>(qres, "OptDQDictMatMulCWu"), std::move(callback));
+}
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:366]: since asym vocabs are now converted to fp16, this can be removed

**Review by dmatveev — COMMENTED** [review]
Self-review

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp (hunk 1 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1578]: it is required, but the comment needs to be updated

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 4]
```diff
@@ -425,10 +472,6 @@ DQParMMGQ::DQParMMGQ(Context::Ref ctx) {
 }
 
 void mergeParallelMatMuls(const std::shared_ptr<ov::Model>& m, Context& ctx) {
-    ov::pass::GraphRewrite rewr;
-    rewr.add_matcher<ov::npuw::patterns::opt::DQParMMGQ>(std::ref(ctx));
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:429]: Does it still work for the repeated block?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 5]
```diff
@@ -208,9 +255,6 @@ DQMatMulGQi::DQMatMulGQi(Context::Ref ctx) {
             matched_qweight->set_partial_shape(tw_shape);
             matched_qweight->validate_and_infer_types();
 
-            matched_qcoeff->set_element_type(ov::element::f16);
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:211]: Why did you move it to `to_f16`? I wonder if those rather invasive changes still work as expected for the repeated block?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp [hunk 6]
```diff
@@ -1596,6 +1640,59 @@ void Partitioner::optimize(const std::string& func_name) {
...
+
+                const auto& gti = ov::get_tensor_impl;
+                if (cw && cz && cs) {
+                    ov::npuw::util::unpack(gti(cw), gti(cz), gti(cs), gti(dst));
+                } else if (cw && cs) {
+                    ov::npuw::util::unpack(gti(cw), gti(cs), gti(dst));
+                } else {
+                    NPUW_ASSERT(false && "Unsupported combination");
+                }
+                funcall._closure.push_back(std::move(dst));
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1677]: Here I go with bank update yet again :)

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 7]
```diff
@@ -516,6 +559,508 @@ void mergeParallelMatMuls(const std::shared_ptr<ov::Model>& m, Context& ctx) {
...
+    };
+    register_matcher(std::make_shared<opp::Matcher>(qgthrw, "HostGather"), std::move(callback));
+}
+
+// Identify the case* where the gather is applied on a compressed
+// (symmetric) vocab tensor. Both CW and GQ paths are supported.
+//
+// FIXME: This may be inefficient: 4x-es the memory consumption
+// due to i4-to-fp16 conversion.
+HostGatherDQ::HostGatherDQ(Context::Ref ctx) {
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:852]: It would be RAM and not NPU memory, right?

**Review by smirnov-alexey — APPROVED** [review]

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp (hunk 6 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:1677]: Yep.. not sure where you do this in your PR but please beware of these changes.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 5 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:211]: As you can see a couple of lines before, `to_f16` looks like some in-place operation rather than something pure. We have
```c++
ctx.get().to_f16(matched_qcoeff);
```
instead of
```c++
auto new_qcoeff = ctx.get().to_f16(matched_qcoeff);
```

So since we mark the original `matched_qcoeff` parameter to be converted, why don't the `to_f16()` do the change to the parameter element type as well? It must've been done there initially, I only figured it out later.

Don't get your point on invasive changes and repeating blocks.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 4 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:429]: It does, why it shouldn't? Though this PR reorganized the order for the passes, this DQParMMGQ is still called in `optimize()` - https://github.com/openvinotoolkit/openvino/pull/26633/files#diff-74c8442fb8280a82e1304119bbf6c1ce6ad4db3980d16b24bef9aa0c41f5fc3aR1603

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 7 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:852]: it all goes to L0 memory.

**Review by smirnov-alexey — APPROVED** [review]
Thanks Dmitry, looks great!
